File: src/newbro/executors/ui/login_item.py
from __future__ import annotations

import plistlib
from pathlib import Path

LOGIN_ITEM_LABEL = "com.newbro.executor-ui"

_LAUNCH_AGENTS_DIR = Path.home() / "Library" / "LaunchAgents"


def login_item_plist_path() -> Path:
    return _LAUNCH_AGENTS_DIR / f"{LOGIN_ITEM_LABEL}.plist"


def render_login_item_plist(*, app_path: Path) -> str:
    payload = {
        "Label": LOGIN_ITEM_LABEL,
        "ProgramArguments": ["/usr/bin/open", str(app_path)],
        "RunAtLoad": True,
    }
    return plistlib.dumps(payload).decode("utf-8")


class LoginItem:
    def __init__(self, *, plist_path: Path | None = None, app_path: Path) -> None:
        self._plist_path = plist_path or login_item_plist_path()
        self._app_path = app_path

    def is_installed(self) -> bool:
        return self._plist_path.exists()

    def install(self) -> None:
        self._plist_path.parent.mkdir(parents=True, exist_ok=True)
        self._plist_path.write_text(render_login_item_plist(app_path=self._app_path), encoding="utf-8")

    def remove(self) -> None:
        self._plist_path.unlink(missing_ok=True)

File: executor-apps/macos/Sources/NewbroExecutorCore/LoginItem.swift
import Foundation

public struct LoginItem {
    public static let label = "com.newbro.executor-ui"

    private let plistPath: URL
    private let appPath: String

    public init(plistPath: URL? = nil, appPath: String) {
        self.plistPath = plistPath ?? FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/LaunchAgents/\(LoginItem.label).plist")
        self.appPath = appPath
    }

    public func renderPlist() -> String {
        let dict: [String: Any] = [
            "Label": LoginItem.label,
            "ProgramArguments": ["/usr/bin/open", appPath],
            "RunAtLoad": true,
        ]
        let data = (try? PropertyListSerialization.data(
            fromPropertyList: dict, format: .xml, options: 0)) ?? Data()
        return String(data: data, encoding: .utf8) ?? ""
    }

    public var isInstalled: Bool {
        FileManager.default.fileExists(atPath: plistPath.path)
    }

    public func install() throws {
        let plist = renderPlist()
        guard !plist.isEmpty, let data = plist.data(using: .utf8) else {
            throw CocoaError(.propertyListWriteStream)
        }
        try FileManager.default.createDirectory(
            at: plistPath.deletingLastPathComponent(), withIntermediateDirectories: true)
        try data.write(to: plistPath)
    }

    public func remove() throws {
        if isInstalled { try FileManager.default.removeItem(at: plistPath) }
    }
}
